### PR TITLE
Ps 5.6 myr merge october2016

### PR DESCRIPTION
**Pull request by george-lorch**

### CODE REVIEW

**Comment by george-lorch** [comment]
Merge October 2016 MyRocks changes into Percona Server 5.6

http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1629/

File: mysql-test/suite/rocksdb/t/add_index_inplace_crash.test [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+--exec echo "restart" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+SET SESSION debug="+d,crash_during_online_index_creation";
+--error 2013
+ALTER TABLE t1 ADD INDEX kb(b), ALGORITHM=INPLACE;
+
+--enable_reconnect
+--source include/wait_until_connected_again.inc
+
+SET SESSION debug="-d,crash_during_online_index_creation";
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/add_index_inplace_crash.test:28]: Redundant (the server, with the previous session debug value, just crashed and restarted with the default)

File: mysql-test/suite/rocksdb/t/add_index_inplace_crash.test [hunk 2]
```diff
@@ -0,0 +1,117 @@
+--source include/have_rocksdb_as_default.inc
+--source include/have_debug.inc
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/add_index_inplace_crash.test:2]: --source include/not_embedded.inc
--source include/not_crashrep.inc

File: mysql-test/suite/rocksdb/t/add_index_inplace_crash.test [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+--exec echo "restart" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+SET SESSION debug="+d,crash_during_index_creation_partition";
+--error 2013
+ALTER TABLE t1 ADD INDEX kij(i,j), ALGORITHM=INPLACE;
+
+--enable_reconnect
+--source include/wait_until_connected_again.inc
+
+SET SESSION debug="-d,crash_during_index_creation_partition";
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/add_index_inplace_crash.test:61]: Likewise

File: mysql-test/suite/rocksdb/t/add_index_inplace_crash.test [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+--enable_query_log
+
+--echo # crash_during_index_creation_partition
+flush logs;
+
+--exec echo "restart" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+SET SESSION debug="+d,myrocks_simulate_index_create_rollback";
+
+--echo # expected assertion failure from sql layer here for alter rollback
+call mtr.add_suppression("Assertion `0' failed.");
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/add_index_inplace_crash.test:96]: Very suspicious suppressions, no other crash testcase does that, shouldn't be needed

File: mysql-test/suite/rocksdb/t/add_index_inplace_crash.test [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+call mtr.add_suppression("Attempting backtrace. You can use the following information to find out");
+
+--error 2013
+
+ALTER TABLE t1 ADD INDEX kij(i,j), ALGORITHM=INPLACE;
+
+--enable_reconnect
+--source include/wait_until_connected_again.inc
+
+SET SESSION debug="-d,myrocks_simulate_index_create_rollback";
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/add_index_inplace_crash.test:106]: See above

File: mysql-test/suite/rocksdb/t/lock_info.test [hunk 6]
```diff
@@ -0,0 +1,33 @@
+--source include/have_rocksdb_as_default.inc
+
+--disable_warnings
+DROP TABLE IF EXISTS t1;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/lock_info.test]: Not needed

File: mysql-test/suite/rocksdb/t/rocksdb.test [hunk 7]
```diff
@@ -818,13 +818,17 @@ update t49 set a = 100 where pk = 1;
...
 set @var1= to_seconds(now());
 send update t49 set a = 1000 where pk = 1;
 
 --connect (con2,localhost,root,,)
 --echo kill query \$con1_id;
 --disable_query_log
+# If we immeditely kill the query - internally the condition broadcast can
+# occur before the lock is waiting on the condition, thus the broadcast call
+# is lost.  Sleep 1 second to avoid this condition.
+--sleep 1
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/rocksdb.test:831]: This sync is going to fail on our Jenkins at once - needs proper rewrite

File: mysql-test/suite/rocksdb/t/rocksdb_deadlock_detect.inc [hunk 8]
```diff
@@ -0,0 +1,84 @@
+#
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/rocksdb_deadlock_detect.inc:1]: Why some of the include files go to t/ and some to include/? They should all go to the latter consistently

File: mysql-test/suite/rocksdb/t/rocksdb_deadlock_detect.inc [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+# Some basic sanity tests for deadlock detection.
+#
+--source include/have_rocksdb_as_default.inc
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/rocksdb_deadlock_detect.inc:4]: --source include/count_sessions.inc

File: mysql-test/suite/rocksdb/t/rocksdb_deadlock_detect.inc [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+select * from t;
+
+connection default;
+disconnect con1;
+disconnect con2;
+disconnect con3;
+
+set global rocksdb_lock_wait_timeout = @prior_rocksdb_lock_wait_timeout;
+set global rocksdb_deadlock_detect = @prior_rocksdb_deadlock_detect;
+drop table t,r1,r2;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/rocksdb_deadlock_detect.inc:84]: --source include/wait_until_count_sessions.inc

File: mysql-test/suite/rocksdb/t/show_engine.test [hunk 11]
```diff
@@ -68,7 +68,8 @@ SHOW ENGINE rocksdb TRANSACTION STATUS;
 
 START TRANSACTION WITH CONSISTENT SNAPSHOT;
 
---replace_regex /[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}/TIMESTAMP/ /(ACTIVE) [0-9]+ /\1 NUM / /(thread id) [0-9]+/\1 TID/ /0x[0-9a-f]+/PTR/
+#select sleep(10);
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/show_engine.test:71]: Remove

File: storage/rocksdb/ha_rocksdb.cc [hunk 12]
```diff
@@ -3824,11 +3862,14 @@ ha_rocksdb::ha_rocksdb(my_core::handlerton *hton,
...
     m_sk_packed_tuple_old(nullptr),
     m_pack_buffer(nullptr), m_lock_rows(RDB_LOCK_NONE),
-    m_keyread_only(FALSE), m_encoder_arr(nullptr),
+    m_keyread_only(FALSE), m_bulk_load_tx(nullptr), m_encoder_arr(nullptr),
     m_row_checksums_checked(0),
     m_in_rpl_delete_rows(false), m_in_rpl_update_rows(false),
     m_force_skip_unique_check(false)
-{}
+{
+  // TODO(alexyang): create a valid PSI_mutex_key for this mutex
```
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc:3870]: Let's put this TODO on their github issues? Mutex instrumentation is a good idea, very easy to implement (not sure why this was not done originally), TODOs in the code are easy to miss

File: storage/rocksdb/rdb_index_merge.cc [hunk 13]
```diff
@@ -478,10 +492,22 @@ int Rdb_index_merge::merge_heap_entry::read_rec(rocksdb::Slice *key,
                                                 rocksdb::Slice *val)
 {
   const uchar* block_ptr= block;
+  const auto orig_offset = chunk_info->curr_offset;
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_index_merge.cc:495]: Inconsistent code style (space before operator)

File: storage/rocksdb/rdb_sst_info.cc [hunk 14]
```diff
@@ -101,22 +110,73 @@ rocksdb::Status Rdb_sst_file::put(const rocksdb::Slice& key,
...
+                            fileinfo.file_size, fileinfo.num_entries);
+    }
+
     // Add the file to the database
-    // Set the skip_snapshot_check parameter to true since no one
+    // Set the snapshot_consistency parameter to false since no one
     // should be accessing the table we are bulk loading
-    s= m_db->AddFile(m_cf, files, true, true);
+    rocksdb::IngestExternalFileOptions opts;
+    opts.move_files = true;
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_sst_info.cc:167]: Likewise

File: storage/rocksdb/rdb_sst_info.h [hunk 15]
```diff
@@ -41,10 +42,14 @@ class Rdb_sst_file {
   const rocksdb::DBOptions&    m_db_options;
   rocksdb::SstFileWriter*      m_sst_file_writer;
   std::string                  m_name;
+  bool                         m_tracing;
+
+  std::string generateKey(const std::string& key);
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_sst_info.h:47]: static

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/suite/rocksdb/t/lock_info.test (hunk 6 above)
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/lock_info.test]: And test should be renamed i_s_locks.

**Comment by george-lorch** [comment]
Alright, well, thanks for the reviews. IMHO all of these I should send upstream as patches. If I change all of this locally, future merging is going to be even more painful. Some I can easily add to my merge patch check script that flags our changed areas so I can easily pick out where I need reconciliation and why. Things like file renames fall into this area. The rest, like code formatting and redundant or unneeded mtr code I would prefer to have it noted here in this PR for reference like you have done, then just accept these as is and later go and send correctional patches to FB and (hopefully) merge the corrections in from FB.

**Pull request merged**